### PR TITLE
Implement TextBox Description Property

### LINKED ISSUE #7381: Add support for `TextBox.Description`
### Current behavior

`TextBox` property `Description` does nothing
image Screenshot_1635024722

### Expected behavior

If `Description` property is set, then it is rendered under `TextBox` via `TextBlock` within a template
image

### How to reproduce it (as minimally and precisely as possible)

1. Create new project
2. Add `TextBox` to `MainPage`
3. Set `Description` property
4. Run

### Workaround

_No response_

### Works on UWP/WinUI

Yes

### Environment

Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia

### NuGet package version(s)

3.10.*

### Affected platforms

iOS, Android, WebAssembly, macOS, Skia (WPF), Skia (GTK on Linux/macOS/Windows), Skia (Tizen)

### IDE

_No response_

### IDE version

17.0 preview 5.0

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by duke7553**
GitHub Issue (If applicable): closes #7381 

## PR Type

What kind of change does this PR introduce?
- Feature

## What is the current behavior?
There is an inconsistency between our TextBox Generic.xaml and that of the Windows platform. Description is not supported.

## What is the new behavior?
Sync with the latest Style for TextBox and add the necessary codebehind. Also, try to fix inconsistent (mixed) formatting of TextBox.cs and Generic.xaml

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information
Tested on WASM. The markup changes should work fine elsewhere, but I'll be happy to test Android too if needed.

Internal Issue (If applicable): N/A

image

### CODE REVIEW

**Comment by davidjohnoliver** [comment]
Thanks for the contribution Luke!

The Uno repo uses tabs for .cs and .xaml - could you fix the whitespace and squash your changes?

If your IDE is set up to consume [.editorconfig files](https://github.com/unoplatform/uno/blob/master/.editorconfig) then the auto-formatting should take care of everything for you.

**Comment by duke7553** [comment]
@davidjohnoliver Thanks for guidance here. I've reverted the formatting mistake the best I could, but please let me know if things are still incorrect and I'll meld another fix into the single commit.

File: src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs [hunk 1]
```diff
@@ -296,6 +296,52 @@ private object CoerceText(object baseValue)
...
+				"Description",
+				typeof(object),
+				typeof(TextBox),
+				new FrameworkPropertyMetadata(
+					defaultValue: null,
+					propertyChangedCallback: (s, e) => ((TextBox)s)?.OnDescriptionChanged(e)
+				)
+			);
+
+		protected void OnDescriptionChanged(DependencyPropertyChangedEventArgs args)
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs:321]: Can be `private`?

**Comment by davidjohnoliver** [comment]
> @davidjohnoliver Thanks for guidance here. I've reverted the formatting mistake the best I could, but please let me know if things are still incorrect and I'll meld another fix into the single commit.

TextBox.cs looks better but some parts were missed, probably inside #CONDITIONAL blocks that weren't active on your solution filter?

Generic.xaml still needs to be fixed.

FYI and in case it helps, here's how I'd personally go about fixing this:
1. Do a soft (technically, mixed) reset with `git reset HEAD^` to pop the commit but keep the file changes
2. View the modified files in a GUI tool that shows whitespace - I like SourceTree for this
3. Check each modified 'hunk' and discard those that don't relate to my intended change - SourceTree again lets you stage/discard granular parts of a modified file
4. Re-commit with fixed-up changes

File: src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs (hunk 1 above)
Comment (duke7553) [reply src/Uno.UI/UI/Xaml/Controls/TextBox/TextBox.cs]: @Youssef1313 Thanks!

File: src/Uno.UI/UI/Xaml/Style/Generic/Generic.xaml [hunk 2]
```diff
@@ -218,38 +218,40 @@
 	</xamarin:Style>
 
 	<!--Default style for Windows.UI.Xaml.Controls.TextBox-->
+	<Thickness x:Key="TextBoxTopHeaderMargin">0,0,0,4</Thickness>
+
 	<Style x:Key="XamlDefaultTextBox"
-				   TargetType="TextBox">
+           TargetType="TextBox">
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/Style/Generic/Generic.xaml:224]: much better but there's still some spaces infiltrating

Comment (duke7553) [reply src/Uno.UI/UI/Xaml/Style/Generic/Generic.xaml:224]: @davidjohnoliver Maybe I'm doing something wrong. How can I remove all space indentations from the file? Is the button in Visual Studio sufficient? I'll try this later today

Comment (duke7553) [reply src/Uno.UI/UI/Xaml/Style/Generic/Generic.xaml]: @davidjohnoliver This should be fixed completely now

Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/Style/Generic/Generic.xaml]: @duke7553 The contents of the `Grid` seem to've swapped positions with the `VisualStateManager`, inside the template. Could you put them back in the original order, the contents of the Grid below the VSM?

Comment (duke7553) [reply src/Uno.UI/UI/Xaml/Style/Generic/Generic.xaml]: @davidjohnoliver Corrected. Let's try again 😆

**Review by davidjohnoliver — APPROVED** [review]

**Comment by davidjohnoliver** [comment]
Ok I think the formatting issues are all out of the way!

The CI is failing with a genuine error though:

> src\Uno.UI\UI\Xaml\Controls\TextBox\TextBox.cs(302,17): Error CS0114: 'TextBox.Description' hides inherited member 'NSObject.Description'. To make the current member override that implementation, add the override keyword. Otherwise add the new keyword.

The problem is that on certain platforms (iOS and macOS), there's already a `Description` property on the base type.

You'll need to add something like:

```csharp
public
#if __IOS__ || __MACOS__
new
#endif
object Description 
...
```

**Comment by duke7553** [comment]
@davidjohnoliver Thank goodness for the pipeline.

**Pull request merged**